src/hdmf/common/resources.py

```diff
@@ -654,3 +708,136 @@ def export_to_sqlite(self, db_file):
             self.entities[:])
         connection.commit()
         connection.close()
+
+    @docval({'name': 'path', 'type': str, 'doc': 'path of the tsv file to write'})
+    def to_tsv(self, **kwargs):
+        """
+        Write ExternalResources as a single, flat table to TSV
+        Internally, the function uses :py:meth:`pandas.DataFrame.to_csv`. Pandas can
+        infer compression based on the filename, i.e., by changing the file extension to
+        ‘.gz’, ‘.bz2’, ‘.zip’, ‘.xz’, or ‘.zst’ we can write compressed files.
+        The TSV is formatted as follows: 1) line one indicates for each column the name of the table
+        the column belongs to, 2) line two is the name of the column within the table, 3) subsequent
+        lines are each a row in the flattened ExternalResources table. The first column is the
+        row id in the flattened table and does not have a label, i.e., the first and second
+        row will start with a tab character, and subseqent rows are numbered sequentially 1,2,3,... .
+        For example:
+
+        .. code-block::
+            :linenos:
+
+            \tobjects\tobjects\tobjects\tobjects\tkeys\tkeys\tresources\tresources\tresources\tentities\tentities\tentities
+            \tobjects_idx\tobject_id\trelative_path\tfield\tkeys_idx\tkey\tresources_idx\tresource\tresource_uri\tentities_idx\tentity_id\tentity_uri
+            0\t0\t1fc87200-e91e-45b3-978c-6d295af144c3\t\tspecies\t0\tMus musculus\t0\tNCBI_Taxonomy\thttps://www.ncbi.nlm.nih.gov/taxonomy\t0\tNCBI:txid10090\thttps://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=10090
+            1\t0\t9bf0c58e-09dc-4457-a652-94065b112c41\t\tspecies\t1\tHomo sapiens\t0\tNCBI_Taxonomy\thttps://www.ncbi.nlm.nih.gov/taxonomy\t1\tNCBI:txid9606\thttps://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=9606
+
+        See also :py:meth:`~hdmf.common.resources.ExternalResources.from_tsv`
+        """  # noqa: E501
+        path = popargs('path', kwargs)
+        df = self.to_dataframe(use_categories=True)
+        df.to_csv(path, sep='\t')
+
+    @classmethod
+    @docval({'name': 'path', 'type': str, 'doc': 'path of the tsv file to read'},
+            returns="ExternalResources loaded from TSV", rtype="ExternalResources")
+    def from_tsv(cls, **kwargs):
+        """
+        Read ExternalResources from a flat tsv file
+        Fromatting of the TSV files are assumed to be consistent with the format
+        generated by :py:meth:`~hdmf.common.resources.ExternalResources.to_tsv`.
+        The function attempts to validate that the data in the TSV is consistent.
+        Currently the checks focus on ensuring that row id links between tables are valid.
```

Comment:
Validating consistency of all values during read may be too expensive, but I think it would be useful to add ``validate_tsv`` function to allows us to validate tsv files. In this way, if a user does indeed edit an ExternalResources table in TSV they can validate it. Also this would make it easier to add validation of ExternalResources the validator. This does not need to be part of this PR, but I just wanted to mention that this would be useful to have. Maybe we can create a separate issue for this.

Reply:
Yes the validation is the next step after this is merged. This is mostly to get ER updated on dev.